x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/type/DateEsField.java

```diff
@@ -6,12 +6,18 @@
  */
 package org.elasticsearch.xpack.esql.core.type;
 
+import org.elasticsearch.common.io.stream.NamedWriteableRegistry;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import java.io.IOException;
 import java.util.Map;
 
 /**
  * SQL-related information about an index field with date type
  */
 public class DateEsField extends EsField {
+    static final NamedWriteableRegistry.Entry ENTRY = new NamedWriteableRegistry.Entry(EsField.class, "DateEsField", DateEsField::new);
```

Comment:
Yeah, fair. There isn't a big standard in the ES code base on how to organize this stuff, but we do tend to do things kind of like I did. But it varies.

As to your first point, it's not entirely true. If you want to deserialize a `DateEsField` directly you can do that - just call it's deserialization ctor. That's private now because no one wants to, but nothing stops us from making it public if you have a need. Some other bits of ESQL will work like this, but I don't believe any `EsField` subclasses will.

The reason for putting the ENTRY in the class itself is:
1. It allows you to make the deserialization ctor private *if* no one else calls it. That's a nice signal.
2. We'll end up with dozens of these entries and it feels good to group them.
3. It's a useful signal to say "this is the category to which this NamedWriteable belongs".

It's also important that the ENTRY ends up near-ish to the class itself so we can test deserialization using it. In this one we register only the NamedWriteables for subclasses of EsField in the test. That's generally a good thing.

Generally the rest of ES seems split. Lots of things do this, like `Block` and `SortValue`. Other things push the entry building up, like to the equivalent of `EsqlPlugin`. Aggs do that and it *fine*. It makes testing serialization a little more complex.

It's a choice and for things shaped like `EsField` I come down on the "put the ENTRY in the class an make a list somewhere convenient" side.

As to leaking the subclasses, I'm not a fan of it, but I also don't hate it too much. Not badly enough to make a new class just to build the entries. The real funny thing about this hierarchy is that the "top" class isn't abstract or an interface or something. Like `Block` is. That bothers me more than the leaking.

Reply:
Fair enough, making serialization tests simple and not involve building a huge registry is nice, indeed.